workflow/executor/executor.go

```diff
@@ -931,6 +958,100 @@ func (we *WorkflowExecutor) Wait(ctx context.Context) error {
 	return nil
 }
 
+var progressRegexp = regexp.MustCompile("#argo .*progress=([0-9]+/[0-9]+)")
+
+// monitorProgress monitors for self-reported progress in the progressFile and patches the pod annotations with the parsed progress.
+//
+// The function reads the last line of the `progressFile` every `readFileTickDuration`.
+// If the line matches `#argo progress=N/M`, will set the progress annotation to the parsed progress value.
+// Every `annotationPatchTickDuration` the pod is patched with the updated annotations. This way the controller
+// gets notified of new self reported progress.
+func (we *WorkflowExecutor) monitorProgress(ctx context.Context, progressFile string) {
+	annotationPatchTicker := time.NewTicker(we.annotationPatchTickDuration)
+	defer annotationPatchTicker.Stop()
+	fileTicker := time.NewTicker(we.readProgressFileTickDuration)
+	defer fileTicker.Stop()
+
+	lastLine := ""
+	progressFile = filepath.Clean(progressFile)
+	// if progress has not been detected once within 10 ticks, abort monitoring
+	countDown := 10
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.WithError(ctx.Err()).Info("context done")
+			return
+		case <-annotationPatchTicker.C:
+			log.WithField("annotations", we.annotations).Infof("patching pod annotations")
+			err := we.patchPodAnnotations(ctx)
+			if err != nil {
+				log.WithError(err).Warn("failed to patch progress annotations")
+			}
+		case <-fileTicker.C:
+			if !we.isRunning(ctx) {
+				log.Info("pod not running yet, skipping tick")
+				continue
+			}
+
+			if countDown == 0 {
+				log.Info("progress not appeared after 10 ticks, cancelling monitoring")
+				return
+			}
+
+			data, err := ioutil.ReadFile(progressFile)
```

Comment:
I think we should store in memory the last value, and check against that before updating

Reply:
not sure I get what you mean - I'm comparing the last line with the most recent line already a few lines below. 
Do you mean to check for the modified date of the file instead (before reading the file)?

Reply:
ah, interesting, so you're assuming the user will append to the file, I'm assuming they'll replace it

I think we should support both, because both are useful

to support replace, you'll need to keep the last value in memory